protocol/block-safety-index.md

```diff
@@ -0,0 +1,224 @@
+to determine which block proposal has to be bound to which L1 blockhash, for reorg safety of L2 proposals.
+
+### The safety indexing problem
+
+Safety indexing is a feature of:
+- explorers: show that a L2 tx became safe at a specific point in L1, possibly even linking the batch data.
+- bridges: determine what L2 blocks to accept for bridging, given some coupling to L1 batch confirmation.
+
+An index, of which L2 block was derived from which L1 block, would help standardize this data,
+and enable every bridge/explorer to interact safely with cross-L2 block data.
+
+# Proposed Solution
+
+Interop, or technically "Isthmus" really, is an upgrade meant to apply to all OP-Stack chains.
+Chains with a dependency-set of 1 (themselves) should not have to run the op-supervisor,
+but should maintain a reliable view of the events and safety of their local chain.
+
+Like the op-supervisor events DB, the history is append-only, except for reorgs.
+This means we could optimize, and store things in an append-only log style system.
+
+The "safedb" in op-node can be replaced with a DB like the above, shared code with the op-supervisor.
+Once the above is done, we can remove the finality mechanism: the op-supervisor signals are used for finalization.
+
+The supervisor is Go, and modular enough that we can embed the single-DB solution into the op-node,
+to serve those less-standard chains with an interop-set of 1.
+This prevents these chains from entering a divergent hardfork path:
+very important to keep the code-base unified and tech-debt low.
+
+The main remaining question is how we bootstrap the data:
+- Chain-sync is on-chain information only, since it needs to verify retrieved data against authenticated commitments.
+  We cannot start from snap-sync if we still have to backfill the history data.
+- Nodes don't currently have this state (if they have not been running the safe-db),
+  we need to roll it out gradually.
+
+This feature could either be a "soft-rollout" type of thing,
+or a Holocene feature (if it turns out to be tightly integrated into the steady batch dervation work).
+
+## Resource Usage
+
+The op-node "safeDB" is closest to what we have today,
+to an index that maps between L2 blocks and L1 blocks that were derived from.
+
+This index is missing space for interop-compatibility (storing both local-safe and cross-safe increments),
+but gives an idea of the resource usage.
+
+In terms of compute it's very minimal to maintain:
+the op-node verifies L1 anyway, and writing the incremental updates is not expensive.
+
+In terms of disk-space it is a little more significant:
+at 32 bytes per L1 and L2 block hash, some L1 metadata, some L2 metadata, and both local-safe cross-safe,
+each entry may about 200 bytes. It does not compress well, as it largely contains blockhash data.
+
+Storing a week of this data, at 2 second blocks, would thus be `7 * 24 * 60 * 60 / 2 * 200 = 60,480,000`, or about 60 MB.
+Storing a year of this data would be around ~3 GB.
```

Comment:
Explorers / indexers might want extended data, to show past L1<>L2 relations. But yes, 1 week should otherwise be enough. Perhaps we can add an archive-flag, where it writes the data that gets pruned to a separate DB, for archival purposes.

Reply:
With clock extension games can be in progress for longer than a week (16 days I think is worst case but may have that wrong) and you typically want to monitor them for some period after they resolve so you have a chance to detect if they resolved incorrectly.  Currently dispute-mon and challenger have a default window of 28 days that they monitor. So I'd suggest we want to keep data for at least that long by default.